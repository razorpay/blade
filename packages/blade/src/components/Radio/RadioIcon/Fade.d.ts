import type { CSSObject } from 'styled-components';

export { Fade } from './Fade.web';

export type FadeProps = {
  show?: boolean;
  children: React.ReactNode;
  styles?: React.CSSProperties | CSSObject;
};
