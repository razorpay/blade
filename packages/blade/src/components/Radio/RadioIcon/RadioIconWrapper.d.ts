export { RadioIconWrapper } from './RadioIconWrapper.web';
