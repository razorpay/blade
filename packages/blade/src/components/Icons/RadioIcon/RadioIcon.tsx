import { Svg, Path } from '../_Svg';
import type { IconComponent } from '..';
import useIconProps from '../useIconProps';

const RadioIcon: IconComponent = ({ size, color, ...styledProps }) => {
  const { height, width, iconColor } = useIconProps({ size, color });

  return (
    <Svg {...styledProps} width={width} height={height} viewBox="0 0 24 24" fill="none">
      <Path
        fillRule="evenodd"
        clipRule="evenodd"
        d="M5.63715 5.63718C6.02762 5.24659 6.02752 4.61343 5.63694 4.22296C5.24636 3.8325 4.61319 3.83259 4.22273 4.22318C-0.071475 8.51867 -0.071475 15.4817 4.22273 19.7772C4.61319 20.1678 5.24636 20.1679 5.63694 19.7774C6.02752 19.3869 6.02762 18.7538 5.63715 18.3632C2.12372 14.8487 2.12372 9.15167 5.63715 5.63718ZM19.7772 4.22318C19.3867 3.83259 18.7535 3.8325 18.3629 4.22296C17.9724 4.61343 17.9723 5.24659 18.3627 5.63718C21.8762 9.15167 21.8762 14.8487 18.3627 18.3632C17.9723 18.7538 17.9724 19.3869 18.3629 19.7774C18.7535 20.1679 19.3867 20.1678 19.7772 19.7772C24.0714 15.4817 24.0714 8.51867 19.7772 4.22318ZM8.46665 8.45768C8.85739 8.06737 8.85775 7.4342 8.46744 7.04346C8.07713 6.65272 7.44397 6.65237 7.05323 7.04268C5.73879 8.35565 5.00022 10.1373 5.00022 11.9952C5.00022 13.853 5.73879 15.6347 7.05323 16.9477C7.44397 17.338 8.07713 17.3376 8.46744 16.9469C8.85775 16.5561 8.85739 15.923 8.46665 15.5327C7.52777 14.5948 7.00022 13.3222 7.00022 11.9952C7.00022 10.6681 7.52777 9.39552 8.46665 8.45768ZM16.9467 7.05268C16.5559 6.66237 15.9227 6.66272 15.5324 7.05346C15.1421 7.4442 15.1425 8.07737 15.5332 8.46768C16.4721 9.40552 16.9997 10.6781 16.9997 12.0052C16.9997 13.3322 16.4721 14.6048 15.5332 15.5427C15.1425 15.933 15.1421 16.5661 15.5324 16.9569C15.9227 17.3476 16.5559 17.348 16.9467 16.9577C18.2611 15.6447 18.9997 13.863 18.9997 12.0052C18.9997 10.1473 18.2611 8.36565 16.9467 7.05268ZM12 9.00018C10.3431 9.00018 9 10.3433 9 12.0002C9 13.657 10.3431 15.0002 12 15.0002C13.6569 15.0002 15 13.657 15 12.0002C15 10.3433 13.6569 9.00018 12 9.00018ZM11 12.0002C11 11.4479 11.4477 11.0002 12 11.0002C12.5523 11.0002 13 11.4479 13 12.0002C13 12.5525 12.5523 13.0002 12 13.0002C11.4477 13.0002 11 12.5525 11 12.0002Z"
        fill={iconColor}
      />
    </Svg>
  );
};

export default RadioIcon;
