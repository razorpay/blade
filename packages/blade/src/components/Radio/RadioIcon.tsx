/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import React from 'react';
import { useTheme } from '~components/BladeProvider';
import Svg from '~components/Icons/_Svg';
import { getIn } from '~utils';
import { CheckboxIconWrapper } from '~components/Checkbox/CheckboxIcon/CheckboxIconWrapper';
import Circle from '~components/Icons/_Svg/Circle';

const CheckedIcon = ({ color }: { color: string }) => {
  return (
    <Svg width="16px" height="16px" viewBox="0 0 16 16" fill="none">
      <Circle cx="8" cy="8" r="7.5" fill={color} stroke={color} />
      <Circle cx="8" cy="8" r="4" fill="white" />
    </Svg>
  );
};

export type RadioIconProps = {
  isFocused?: boolean;
  isDisabled?: boolean;
  isNegative?: boolean;
  isChecked?: boolean;
};

const RadioIcon = ({ isChecked, isDisabled, isNegative }: RadioIconProps) => {
  const { theme } = useTheme();
  const defaultIconColor = getIn(theme, 'colors.brand.primary.200');
  const disabledIconColor = getIn(theme, 'colors.brand.gray.500');
  const iconColor = isDisabled ? disabledIconColor : defaultIconColor;

  return (
    <CheckboxIconWrapper
      isFullRounded={true}
      isDisabled={isDisabled}
      isNegative={isNegative}
      isChecked={!!isChecked}
    >
      {isChecked ? <CheckedIcon color={iconColor} /> : null}
    </CheckboxIconWrapper>
  );
};

export { RadioIcon };
