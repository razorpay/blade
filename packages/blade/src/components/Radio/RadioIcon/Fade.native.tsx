/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import React from 'react';
import type { EasingFn } from 'react-native-reanimated';
import Animated, { Keyframe } from 'react-native-reanimated';
import type { CSSObject } from 'styled-components';
import styled from 'styled-components/native';
import type { FadeProps } from './types';
import { useTheme } from '~components/BladeProvider';

const StyledFade = styled(Animated.View)<{ styles: CSSObject }>(({ styles }) => {
  return {
    ...styles,
  };
});

const Fade = ({ children, show, styles }: FadeProps) => {
  const { theme } = useTheme();

  const enterEasing = (theme.motion.easing.entrance.effective as unknown) as EasingFn;
  const exitEasing = (theme.motion.easing.exit.effective as unknown) as EasingFn;
  const scaleIn = new Keyframe({
    from: {
      transform: [{ scale: 0.3 }],
      opacity: 0,
      easing: enterEasing,
    },
    to: {
      transform: [{ scale: 1 }],
      opacity: 1,
      easing: enterEasing,
    },
  }).duration(theme.motion.duration.quick);

  const fadeOut = new Keyframe({
    from: {
      opacity: 1,
      easing: exitEasing,
    },
    to: {
      opacity: 0,
      easing: exitEasing,
    },
  }).duration(theme.motion.duration.quick);

  return show ? (
    <StyledFade styles={styles as CSSObject} entering={scaleIn} exiting={fadeOut}>
      {children}
    </StyledFade>
  ) : null;
};

export { Fade };
