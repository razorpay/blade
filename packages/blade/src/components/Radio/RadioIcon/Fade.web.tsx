/* eslint-disable react/jsx-no-useless-fragment */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import React from 'react';
import type { FlattenSimpleInterpolation } from 'styled-components';
import styled, { css, keyframes } from 'styled-components';
import usePresence from 'use-presence';
import type { FadeProps } from './types';
import { useTheme } from '~components/BladeProvider';
import { makeMotionTime } from '~utils';

const scaleIn = keyframes`
  from {
    transform: scale(.3);
    opacity: 0;
  }

  to {
    transform: scale(1);
    opacity: 1;
  }
`;
const fadeOut = keyframes`
  from {
    transform: scale(1);
    opacity: 1;
  }

  to {
    transform: scale(1);
    opacity: 0;
  }
`;

const AnimatedFade = styled.div<{ animationType: FlattenSimpleInterpolation | null }>(
  ({ animationType }) =>
    animationType === null
      ? ''
      : css`
          ${animationType}
        `,
);

const Fade = ({ show, children, styles }: FadeProps) => {
  const { theme } = useTheme();

  const duration = theme.motion.duration.xquick;
  const enter = css`
    animation: ${scaleIn} ${makeMotionTime(duration)}
      ${theme.motion.easing.entrance.effective as string};
  `;

  const exit = css`
    animation: ${fadeOut} ${makeMotionTime(duration)}
      ${theme.motion.easing.exit.effective as string};
  `;

  // usePresence hook waits for the animation to finish before unmounting the component
  // It's similar to framer-motions usePresence hook
  // https://www.framer.com/docs/animate-presence/#usepresence
  const { isMounted, isVisible } = usePresence(!!show, {
    transitionDuration: duration,
    initialEnter: true,
  });

  return (
    <>
      {isMounted && (
        <AnimatedFade animationType={isVisible ? enter : exit} style={styles}>
          {children}
        </AnimatedFade>
      )}
    </>
  );
};

export { Fade };
