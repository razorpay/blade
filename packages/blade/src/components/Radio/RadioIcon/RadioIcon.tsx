/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import getIn from 'lodash/get';
import React from 'react';
import { radioIconColors, radioSizes } from '../radioTokens';
import type { RadioProps } from '../Radio';
import { RadioIconWrapper } from './RadioIconWrapper';
import { Fade } from './Fade';
import { useTheme } from '~components/BladeProvider';
import Svg from '~components/Icons/_Svg';
import Circle from '~components/Icons/_Svg/Circle';
import { makeSpace } from '~utils/makeSpace';
import { metaAttribute } from '~utils/metaAttribute';

export type RadioIconProps = {
  isDisabled?: boolean;
  isNegative?: boolean;
  isChecked?: boolean;
} & Required<Pick<RadioProps, 'size'>>;

const CheckedIcon = ({ color, size }: { color: string; size: RadioIconProps['size'] }) => {
  const width = radioSizes.icon[size].width;
  const height = radioSizes.icon[size].height;
  const radius = radioSizes.icon[size].dotRadius;
  const viewBox = `0 0 ${width} ${height}`;
  const cx = `${width / 2}`;
  const cy = `${height / 2}`;

  return (
    <Svg width={makeSpace(width)} height={makeSpace(height)} viewBox={viewBox} fill="none">
      <Circle cx={cx} cy={cy} r={`${radius}`} fill={color} />
    </Svg>
  );
};

const RadioIcon = ({ isChecked, isDisabled, isNegative, size }: RadioIconProps) => {
  const { theme } = useTheme();

  const checked = Boolean(isChecked);
  const state = checked ? 'checked' : 'unchecked';
  let variant: 'default' | 'disabled' | 'negative' = 'default';
  if (isDisabled) variant = 'disabled';
  if (isNegative) variant = 'negative';
  const dotColor = getIn(theme, radioIconColors.variants[variant].dot[state]);

  return (
    <RadioIconWrapper
      size={size}
      isDisabled={isDisabled}
      isNegative={isNegative}
      isChecked={checked}
      {...metaAttribute({ name: 'radio-icon-wrapper' })}
    >
      <Fade show={checked} styles={{ position: 'absolute', display: 'flex' }}>
        <CheckedIcon size={size} color={dotColor} />
      </Fade>
    </RadioIconWrapper>
  );
};

export { RadioIcon };
