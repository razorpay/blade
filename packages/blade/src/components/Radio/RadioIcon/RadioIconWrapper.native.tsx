import styled from 'styled-components/native';
import type { RadioRectProps } from './RadioIconWrapperStyles';
import { getRadioIconWrapperStyles } from './RadioIconWrapperStyles';

const RadioIconWrapper = styled.View<RadioRectProps>((props) => getRadioIconWrapperStyles(props));

export { RadioIconWrapper };
