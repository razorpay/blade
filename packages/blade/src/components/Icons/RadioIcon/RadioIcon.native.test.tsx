import RadioIcon from './';
import renderWithTheme from '~utils/testing/renderWithTheme.native';

describe('<RadioIcon />', () => {
  it('should render RadioIcon', () => {
    const renderTree = renderWithTheme(
      <RadioIcon color="feedback.icon.neutral.intense" size="large" />,
    ).toJSON();
    expect(renderTree).toMatchSnapshot();
  });
});
