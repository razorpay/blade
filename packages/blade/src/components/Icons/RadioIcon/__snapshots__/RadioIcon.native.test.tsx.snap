// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioIcon /> should render RadioIcon 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNSVGSvgView
    accessibilityElementsHidden={true}
    align="xMidYMid"
    bbHeight="20px"
    bbWidth="20px"
    data-blade-component="icon"
    fill="none"
    focusable={false}
    height="20px"
    importantForAccessibility="no-hide-descendants"
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        [
          {},
        ],
        {
          "flex": 0,
          "height": 20,
          "width": 20,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width="20px"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
        ]
      }
    >
      <RNSVGPath
        clipRule={0}
        d="M5.63715 5.63693C6.02762 5.24635 6.02752 4.61318 5.63694 4.22272C5.24636 3.83225 4.61319 3.83235 4.22273 4.22293C-0.071475 8.51843 -0.071475 15.4814 4.22273 19.7769C4.61319 20.1675 5.24636 20.1676 5.63694 19.7771C6.02752 19.3867 6.02762 18.7535 5.63715 18.3629C2.12372 14.8484 2.12372 9.15143 5.63715 5.63693ZM19.7772 4.22293C19.3867 3.83235 18.7535 3.83225 18.3629 4.22272C17.9724 4.61318 17.9723 5.24635 18.3627 5.63693C21.8762 9.15143 21.8762 14.8484 18.3627 18.3629C17.9723 18.7535 17.9724 19.3867 18.3629 19.7771C18.7535 20.1676 19.3867 20.1675 19.7772 19.7769C24.0714 15.4814 24.0714 8.51843 19.7772 4.22293ZM8.46665 8.45743C8.85739 8.06712 8.85775 7.43396 8.46744 7.04322C8.07713 6.65248 7.44397 6.65212 7.05323 7.04243C5.73879 8.35541 5.00022 10.1371 5.00022 11.9949C5.00022 13.8528 5.73879 15.6345 7.05323 16.9474C7.44397 17.3377 8.07713 17.3374 8.46744 16.9466C8.85775 16.5559 8.85739 15.9227 8.46665 15.5324C7.52777 14.5946 7.00022 13.322 7.00022 11.9949C7.00022 10.6679 7.52777 9.39527 8.46665 8.45743ZM16.9467 7.05243C16.5559 6.66212 15.9227 6.66248 15.5324 7.05322C15.1421 7.44396 15.1425 8.07712 15.5332 8.46743C16.4721 9.40527 16.9997 10.6779 16.9997 12.0049C16.9997 13.332 16.4721 14.6046 15.5332 15.5424C15.1425 15.9327 15.1421 16.5659 15.5324 16.9566C15.9227 17.3474 16.5559 17.3477 16.9467 16.9574C18.2611 15.6445 18.9997 13.8628 18.9997 12.0049C18.9997 10.1471 18.2611 8.36541 16.9467 7.05243ZM12 8.99994C10.3431 8.99994 9 10.3431 9 11.9999C9 13.6568 10.3431 14.9999 12 14.9999C13.6569 14.9999 15 13.6568 15 11.9999C15 10.3431 13.6569 8.99994 12 8.99994ZM11 11.9999C11 11.4477 11.4477 10.9999 12 10.9999C12.5523 10.9999 13 11.4477 13 11.9999C13 12.5522 12.5523 12.9999 12 12.9999C11.4477 12.9999 11 12.5522 11 11.9999Z"
        fill={
          {
            "payload": 4280563015,
            "type": 0,
          }
        }
        fillRule={0}
        propList={
          [
            "fill",
            "fillRule",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
